frontend/src/modules/progresschart/components/Canvas.js

```diff
@@ -0,0 +1,116 @@
+/* @flow */
+
+import * as React from 'react';
+
+import './Canvas.css';
+
+import type { Stats } from 'core/stats';
+
+
+type Props = {|
+    stats: Stats,
+    height: number,
+    width: number,
+    class: string,
+|};
+
+
+/**
+ * Render current resource progress data on canvas.
+ */
+export default class Canvas extends React.Component<Props> {
+    canvas: { current: HTMLCanvasElement | null };
+
+    constructor(props: Props) {
+        super(props);
+        this.canvas = React.createRef();
+    }
+
+    componentDidMount() {
+        this.setUpCanvas();
+        this.drawCanvas();
+    }
+
+    componentDidUpdate() {
+        this.drawCanvas();
+    }
+
+    setUpCanvas() {
+        const dpr = window.devicePixelRatio || 1;
+        const canvas = this.canvas.current;
+
+        if (!canvas) {
+            return;
+        }
+
+        // Set up canvas to be HiDPI display ready
+        canvas.style.width = canvas.width + 'px';
+        canvas.style.height = canvas.height + 'px';
+        canvas.width = canvas.width * dpr;
+        canvas.height = canvas.height * dpr;
+    }
+
+    drawCanvas() {
+        const { approved, fuzzy, warnings, errors, missing, total } = this.props.stats;
+        const dpr = window.devicePixelRatio || 1;
+        const canvas = this.canvas.current;
+
+        const data = [
+            {
+                type: total ? approved / total : 0,
+                color: '#7BC876',
+            },
+            {
+                type: total ? fuzzy / total : 0,
+                color: '#FED271',
+            },
+            {
+                type: total ? warnings / total : 0,
+                color: '#FFA10F',
+            },
+            {
+                type: total ? errors / total : 0,
+                color: '#F36',
+            },
+            {
+                type: total ? missing / total : 0,
```

Comment:
These are the same as `total && missing / total`. Not sure if that's more readable though, I'll let you judge that.